Mathlib/Order/Directed.lean

```diff
@@ -321,3 +321,65 @@ instance (priority := 100) OrderTop.to_isDirected_le [LE α] [OrderTop α] : IsD
 -- see Note [lower instance priority]
 instance (priority := 100) OrderBot.to_isDirected_ge [LE α] [OrderBot α] : IsDirected α (· ≥ ·) :=
   ⟨fun _ _ => ⟨⊥, bot_le _, bot_le _⟩⟩
+
+namespace DirectedOn
+
+section Pi
+
+namespace Pi
+
+variable {ι : Type*} {α : ι → Type*} [∀ i, LE (α i)]
+
+lemma proj {d : Set (Π i, α i)} (hd : DirectedOn (· ≤ ·) d) (i : ι) :
+    DirectedOn (· ≤ ·) ((fun a => a i) '' d) := by
+  intro p ⟨a, ha⟩ q ⟨b, hb⟩
+  obtain ⟨z,hz⟩ := hd a ha.1 b hb.1
+  use z i
+  aesop
+
+lemma prodMk {d : (i : ι) → Set (α i)} (hd : ∀ (i : ι), DirectedOn (· ≤ ·) (d i)) :
+    DirectedOn (· ≤ ·) (Set.pi  Set.univ d) := by
+  intro a ha b hb
+  dsimp
+  choose f hf using fun i => hd i (a i) (ha i trivial) (b i) (hb i trivial)
+  simpa [Pi.le_def, ← forall_and] using ⟨f, hf⟩
+
+end Pi
+
+end Pi
+
+section Prod
+
+variable {r₂ : β → β → Prop}
+
+/-- Local notation for a relation -/
+local infixl:50 " ≼₁ " => r
+/-- Local notation for a relation -/
+local infixl:50 " ≼₂ " => r₂
```

Comment:
Could you add docstrings to the lemmas below saying explicitly that they are applicable to product orders thanks to definitional equality (if that is indeed the case)?

Also, is there a reason to allow for non-le relations in this case but not for indexed products? My guess is it work as well.

Reply:
The `LE` results can be recovered from the new results as:
```
variable {ι : Type*} {α : ι → Type*} [∀ i, LE (α i)]

example {d : Set (Π i, α i)} (hd : DirectedOn (· ≤ ·) d) (i : ι) :
    DirectedOn (· ≤ ·) ((fun a => a i) '' d) := proj _ hd _

example {d : (i : ι) → Set (α i)} (hd : ∀ (i : ι), DirectedOn (· ≤ ·) (d i)) :
    DirectedOn (· ≤ ·) (Set.pi  Set.univ d) := pi _ (fun i => hd i)
```